src/core/storage/sframe_data/groupby_aggregate_impl.cpp

```diff
@@ -217,128 +220,175 @@ size_t groupby_element::hash() const {
+  if (groupby_element_vec_ptr == NULL)
+    groupby_element_vec_ptr = new std::vector<groupby_element>;
   // note. not auto&. This needs to take a real value (pointer) and not a
   // reference since the auto& will make it really a reference to a pointer to
   // a vector<groupby_element> which will make it not robust to array resizes.
   auto groupby_element_vec = groupby_element_vec_ptr;
-  segments[target_segment].refctr.inc();
-  lock.unlock();
-  segments[target_segment].fine_grain_locks[hash % 128].lock();
-  bool found = false;
-  for (size_t i = 0;i < groupby_element_vec->size(); ++i) {
+
+  for (size_t i = 0; i < groupby_element_vec->size(); ++i) {
     if (flexible_type_vector_equality((*groupby_element_vec)[i].key,
-                                      (*groupby_element_vec)[i].key.size(),
-                                      val,
+                                      (*groupby_element_vec)[i].key.size(), val,
                                       num_keys)) {
       (*groupby_element_vec)[i].add_element(val, group_descriptors);
-      found = true;
-      break;
+      if (segments[target_segment].elements_.size() >= max_buffer_size) {
+        flush_segment(target_segment);
+      }
+      return;
     }
   }
-  if (!found) {
-    std::vector<flexible_type> keys; keys.reserve(num_keys);
-    for (size_t i = 0;i < num_keys; ++i) keys.push_back(val[i]);
-
-    groupby_element_vec->push_back(groupby_element{
-                                     std::move(keys),
-                                     group_descriptors});
-    (*groupby_element_vec)[groupby_element_vec->size() - 1].add_element(val, group_descriptors);
-  }
-  segments[target_segment].fine_grain_locks[hash % 128].unlock();
-  segments[target_segment].refctr.dec();
-  // element not found
-  if (segments[target_segment].elements.size() >= max_buffer_size) {
-    flush_segment(target_segment);
-  }
+
+  std::vector<flexible_type> keys;
+  keys.reserve(num_keys);
+  for (size_t i = 0; i < num_keys; ++i) keys.push_back(val[i]);
+
+  groupby_element_vec->push_back(
+      groupby_element{std::move(keys), group_descriptors});
+
+  (*groupby_element_vec)[groupby_element_vec->size() - 1].add_element(
+      val, group_descriptors);
 }
 
 void group_aggregate_container::flush_segment(size_t segmentid) {
-  // unlock and swap out the segment.
-  std::unique_lock<turi::simple_spinlock> lock(segments[segmentid].in_memory_group_lock);
-  if (segments[segmentid].elements.size() == 0) return;
-  while(segments[segmentid].refctr.value > 0) cpu_relax();
-  decltype(segments[segmentid].elements) local;
-  local.swap(segments[segmentid].elements);
-  lock.unlock();
+  logstream(LOG_DEBUG) << "flush buffer from task_id: " << tss_.id_
+                       << ", segment_id" << segmentid << std::endl;
+
+  auto& segments = tss_.segments_;
+  if (segments[segmentid].elements_.size() == 0) return;
+  // what is cpu_relax
+  // while(segements_[segmentid].refctr.value > 0) cpu_relax();
```

Comment:
If this line is not needed anymore, then let's remove it.  But are we sure it's not needed?

cpu_relax() is a bit like yield(); It tells the cpu to prioritize other threads for a bit.  Otherwise spin-lock type wait loops can be very cpu intensive.

Reply:
oh, got it! I figured it out before and I forget to remove it.